File: app/src/main/java/com/soundcloud/android/facebookinvites/FacebookCreatorInvitesItemRenderer.java
package com.soundcloud.android.facebookinvites;

import com.soundcloud.android.R;
import com.soundcloud.android.events.EventQueue;
import com.soundcloud.android.events.FacebookInvitesEvent;
import com.soundcloud.android.image.ApiImageSize;
import com.soundcloud.android.image.ImageOperations;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.presentation.CellRenderer;
import com.soundcloud.android.stream.StreamItem;
import com.soundcloud.rx.eventbus.EventBus;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import javax.inject.Inject;
import java.util.List;

public class FacebookCreatorInvitesItemRenderer implements CellRenderer<StreamItem.FacebookCreatorInvites> {

    private final ImageOperations imageOperations;
    private final FacebookInvitesStorage facebookInvitesStorage;
    private final EventBus eventBus;
    private Listener listener;

    public interface Listener {
        void onCreatorInvitesDismiss(int position);

        void onCreatorInvitesClicked(int position);
    }

    @Inject
    public FacebookCreatorInvitesItemRenderer(ImageOperations imageOperations,
                                              FacebookInvitesStorage facebookInvitesStorage,
                                              EventBus eventBus) {
        this.imageOperations = imageOperations;
        this.facebookInvitesStorage = facebookInvitesStorage;
        this.eventBus = eventBus;
    }

    @Override
    public View createItemView(ViewGroup parent) {
        eventBus.publish(EventQueue.TRACKING, FacebookInvitesEvent.forCreatorShown());
        return LayoutInflater.from(parent.getContext())
                             .inflate(R.layout.facebook_creator_invites_notification_card, parent, false);
    }

    @Override
    public void bindItemView(int position, View itemView, List<StreamItem.FacebookCreatorInvites> items) {
        ImageView artwork = (ImageView) itemView.findViewById(R.id.artwork);
        Urn trackUrn = items.get(position).trackUrn();
        itemView.setEnabled(false);
        setClickListeners(itemView, position);
        imageOperations.displayWithPlaceholder(trackUrn, ApiImageSize.T300, artwork);
    }

    public void setOnFacebookInvitesClickListener(Listener listener) {
        this.listener = listener;
    }

    private void setClickListeners(View itemView, final int position) {
        itemView.findViewById(R.id.close_button).setOnClickListener(v -> {
            facebookInvitesStorage.setCreatorDismissed();
            if (listener != null) {
                listener.onCreatorInvitesDismiss(position);
            }
        });

        itemView.findViewById(R.id.action_button).setOnClickListener(v -> {
            facebookInvitesStorage.setClicked();
            if (listener != null) {
                listener.onCreatorInvitesClicked(position);
            }
        });
    }
}

File: app/src/main/java/com/soundcloud/android/facebookinvites/FacebookCreatorInvitesItemRenderer.kt
package com.soundcloud.android.facebookinvites

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.soundcloud.android.R
import com.soundcloud.android.events.EventQueue
import com.soundcloud.android.events.FacebookInvitesEvent
import com.soundcloud.android.image.ApiImageSize
import com.soundcloud.android.image.ImageOperations
import com.soundcloud.android.presentation.CellRenderer
import com.soundcloud.android.stream.StreamItem
import com.soundcloud.java.optional.Optional
import com.soundcloud.rx.eventbus.EventBusV2
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.facebook_creator_invites_notification.view.*
import kotlinx.android.synthetic.main.facebook_creator_invites_notification_card.view.*
import javax.inject.Inject

class FacebookCreatorInvitesItemRenderer
@Inject
constructor(private val imageOperations: ImageOperations,
            private val facebookInvitesStorage: FacebookInvitesStorage,
            private val eventBus: EventBusV2) : CellRenderer<StreamItem.FacebookCreatorInvites> {
    val loadingResult: PublishSubject<FacebookLoadingResult> = PublishSubject.create()

    override fun createItemView(parent: ViewGroup): View {
        eventBus.publish(EventQueue.TRACKING, FacebookInvitesEvent.forCreatorShown())
        return LayoutInflater.from(parent.context)
                .inflate(R.layout.facebook_creator_invites_notification_card, parent, false)
    }

    override fun bindItemView(position: Int, itemView: View, items: List<StreamItem.FacebookCreatorInvites>) {
        itemView.isEnabled = false
        itemView.close_button.setOnClickListener {
            facebookInvitesStorage.setCreatorDismissed()
            loadingResult.onNext(FacebookLoadingResult.Dismiss(position))
        }
        itemView.action_button.setOnClickListener {
            facebookInvitesStorage.setClicked()
            loadingResult.onNext(FacebookLoadingResult.Click(position))
        }
        imageOperations.displayWithPlaceholder(
                items[position].trackUrn,
                Optional.absent(),
                ApiImageSize.T300,
                itemView.artwork)
    }

}
